docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+In this case, the transition fails, indicating to the router that is should
+close the newly established connection.
+
+> Question: is it possible to have multiple routines dialing to the same peer?
+> This could be another reason for this transition to fail.
+
+It may also occur that the node is already connected to `MaxConnected` peers,
+which means that all connection slots are full.
+In this case, the peer manager tries to find a connection slot that can be
+[upgraded](#slots-upgrading) to give room for the new established connection.
+If no suitable connection slot is found, or the hard limit `MaxConnected +
+MaxConnectedUpgrade` of connected peers is reached, the transitions fails.
+
+Notice that, in order to dial this peer, the peer manager may have put another
+lower-ranked peer in the [upgrading sub-state](#upgrading-peeer) to give room
+to this connection.
+In this case, the slot for the established connection was *reserved*, and this
+transition will not fail.
+
+If the transition succeeds, the peer is set to the
+[`Connected`](#connected-peer) state as an `outgoing` peer.
+The peer's `LastConnected` and the dialed address' `LastDialSuccess` times are
+set, and dialed address' `DialFailures` counter is reset.
+
+> If the peer is `Inactive`, it is set as active.
+> This action has no effect apart from producing metrics.
+
+If a connection slot was upgraded to give room for the established connection, the
+peer on that slot transitions to the [evict sub-state](#evict-peer).
+
+#### Errors
+
+The transition fails if:
+
+- the node dialed itself
+- the peer is already in the `Connected` state
+- the node is connected to `MaxConnected` peers, and no slot is suitable for upgrading
+
+Errors are also returned if:
+
+- the dialed peer was removed from the peer store
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### DialFailed transition
+
+This transition informs a failure when establishing an `outgoing` connection to
+a peer.
+
+The dialed address's `LastDialFailure` time is set, and its `DialFailures`
+counter is increased.
+This information is used to compute the [retry delay](#retry-delay) for the
+dialed address.
+
+The peer manager then spawns a routine that after the computed retry delay
+notifies the next peer to dial routine about the availability of this peer.
+Until then, the peer is the `Frozen Candidate` state.
+
+#### Retry delay
+
+The retry dial is the minimum time, from the latest failed dialing attempt, we
+should wait until dialing a peer address again.
+
+The default delay is defined by `MinRetryTime` parameter.
+If it is set to zero, we *never* retry dialing a peer address.
+
+Upon each failed dial attempt, we increase the delay by `MinRetryTime`, plus an
+optional random jitter of up to `RetryTimeJitter`.
+
+The retry delay should not be longer than the `MaxRetryTime` parameter,
+or `MaxRetryTimePersistent` parameter in the case of persistent peers.
```

Comment:
This is a good question.

From what I remember from the previous version, we should never give up trying to connect to a persistent peer. This is a peer we should be connected to, and the fact that it is offline for a while should not change this assumption. Since the retry delay is increasing, we can reach a point in which it is very large, meaning that we will not react promptly to the fact that our persistent peer is back online.

For non-persistent peers this reasoning makes less sense, in my opinion.